### DISCUSSION TITLE
IDs used for entities for LDAP stores

### DISCUSSION THREAD

**Discussion by ahus1** [keycloak/keycloak#10121] [Archive]
# Problem statement

The current implementation of the Map storage proceeds as follows for creating a new realm role:

* RealmCacheSession#addRealmRole(RealmModel realm, String name) that will generate an ID for the role as UUID
* other map transactions create the ID when create() is called
* when creating the entity in LDAP, the LDAP will decide about the UUID on creation time, and it can't be passed by the caller. For other stores, writing to the store would usually be done at the end of the transaction
  
# Approach

The first part seems quite clear: Remove code to create ID from RealmCacheSession

Then it gets tricky, here a first approach:

* instead of creating the entity at the end of the transaction, call LDAP from LdapRoleMapKeycloakTransaction#create(MapRoleEntity value) to create an entity and get the GUID that will be used as ID. When the transaction is rolled back, the entity would need to be removed. 

# Benefits

* Using a GUID would not leak DN information outside the LDAP storage provider.

# Alternatives

* Use the DN of the entity that could be determined at the time LdapRoleMapKeycloakTransaction#create is being called. Only during commit the entity is created in LDAP as usual.

Benefit (or downside?) of a DN would be that an entity or a directory could be re-created with the same DNs, while it can't be re-created with the same GUIDs.

# Other

* References inside LDAP (for example "member" and "memberOf" attributes) are stored as DN as well (probably for a reason), and it would be more human readable.

**Comment by ahus1** [comment]
Follow up question: if there are multiple stores involved in a tree store, which store determines the ID of the entity ... and will all stores require to accept an ID being passed to them?

**Comment by hmlnarik** [comment]
Thank you, I agree with the suggested approach. This would basically mean creating an auxliliary object in LDAP with say "cn=GUID,..." DN and upon commit "renaming" it to a proper name, or removing it upon rollback.

Side notes:
- `RealmCacheSession` does applies to legacy store, not to map storage.
- Map storages are free to return a different object with a different ID from `create` than was requested. The ID in the object is a hint. This is not explicitly mentioned in the javadoc (`It may or may not be the same instance as {@code value}`), it could have been phrased better there
- References inside LDAP might be stored as GUIDs and only representation returned is a DN - is there any reference supporting that it is stored as DN internally?

**Reply by ahus1** [reply]
I'll update the docs to state that a map storage might disregard the ID passed by the caller and can replace it something else. Taking the current JavaDoc at face value, I assumed it was only talking about Java object instances. 

From an API point of view it might be simpler when every map storage implementation creates new IDs at its own discretion; that would prevent a situation when someone memorized the ID before by mistake, and only notices that when switching to the one storage provider that changes the ID. 

I tested references to member attributes as part of groupOfNames. The attributes for an Apache DS instance show up in Apache Directory Studio as DNs. They can be change to any different string without Apache DS complaining. So there is no evidence to me that at least Apache DS stores them internally as GUIDs. 

On the other hand I've read that Active Directory updates a virtual [memberOf](https://ldapwiki.com/wiki/MemberOf) attributes once something is a member. Apache DS doesn't do this.
